### PR TITLE
Ship printing costs

### PR DESCRIPTION
**Pull request by SciathGH**

### CODE REVIEW

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt [hunk 1]
```diff
@@ -131,6 +136,30 @@ object BlueprintCommand : SLCommand() {
...
 	}
 
+	private fun calculateBlueprintCost(blueprint: Blueprint) : Double {
+		val clipboard = blueprint.loadClipboard()
+
+		var PriceAmmount = 0.0
+
+		val materialmap = mutableMapOf<BlockData, Int>()
+
+		for (vec in clipboard.region) {
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: clipboard.region.map

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		Material.SEA_LANTERN to 25.0,
+		Material.IRON_TRAPDOOR to 50.0,
+		Material.NOTE_BLOCK to 20.0,
+		Material.IRON_BARS to 10.0,
+		Material.OBSERVER to 5.0,
+		Material.TRAPPED_CHEST to 5.0,
+		Material.BARREL to 5.0
+	)
+	fun getPrice(data: BlockData) : Double {
+		if(data.material.isGlass) return 0.1
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:61]: return if / else if / else ?:

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import net.starlegacy.util.isWallSign
+import org.bukkit.Material
+import org.bukkit.block.data.BlockData
+
+object ShipFactoryMaterialCosts {
+	var blockprice = mapOf<Material, Double>(
+		Material.IRON_BLOCK to 50.0,
+		Material.DIAMOND_BLOCK to 50.0,
+		Material.COPPER_BLOCK to 175.0,
+		CustomBlocks.MINERAL_URANIUM.block.blockData.material to 50.0,
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:21]: the material of these will all be the same

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt [hunk 4]
```diff
@@ -122,7 +125,9 @@ object BlueprintCommand : SLCommand() {
 
 	private fun blueprintInfo(blueprint: Blueprint): List<String> {
 		val list = LinkedList<String>()
+		var blueprintcost = calculateBlueprintCost(blueprint)
 		list.add("&7Size&8: &6${blueprint.size}")
+		list.add("&7Cost&8: &6$${blueprintcost.roundToInt()}")
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:130]: why not just .toInt()

**Review by Peter-Crawley — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt (hunk 1 above)
Comment (SciathGH) [reply src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: cant do return in there

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt (hunk 2 above)
Comment (SciathGH) [reply src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:61]: sure why not

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt (hunk 4 above)
Comment (SciathGH) [reply src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:130]: roundToInt is cleaner, but if u want sure

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt (hunk 2 above)
Comment (trainb0y) [reply src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:61]: return when :iq: /hj

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt (hunk 1 above)
Comment (Peter-Crawley) [reply src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: ? you can map to an array of int and then use another function to add them all up

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt (hunk 2 above)
Comment (SciathGH) [reply src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:61]: not doing "else" it could cause anything not in there to be returned 0.1

Comment (Peter-Crawley) [reply src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:61]: cant use when there

Comment (SciathGH) [reply src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:61]: i can do else if, so im going too

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt (hunk 1 above)
Comment (SciathGH) [reply src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: That sounds like a waste of time, the way I originally have it works fine, and doesnt require any extra effort on anyones part

Comment (Peter-Crawley) [reply src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: cleaner

Comment (SciathGH) [reply src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: Ok, how is it cleaner then 1 for loop?

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		else if(data.material.isConcrete) return 0.1
+		else if(data.material.isGlassPane) return 0.1
+		else if(data.material.isSlab) return 0.1
+		else if(data.material.isStainedTerracotta) return 0.1
+		else if(data.material.isGlazedTerracotta) return 0.1
+		else if(data.material.isConcrete) return 0.1
+		else if(data.material.isSign) return 25.0
+		else if(data.material.isWallSign) return 25.0
+		else if(data.material.isWall) return 0.1
+		when(data){
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:68]: else when

File: src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+			CustomBlocks.MINERAL_TITANIUM.block.blockData -> return 200.0
+			CustomBlocks.MINERAL_URANIUM.block.blockData -> return 50.0
+			CustomBlocks.MINERAL_CHETHERITE.block.blockData -> return 100.0
+			CustomBlocks.MINERAL_ALUMINUM.block.blockData -> return 100.0
+		}
+		return blockprice[data.material] ?: 3.0
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/horizonsend/ion/core/ShipFactoryMaterialCosts.kt:74]: else if

File: src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt [hunk 7]
```diff
@@ -131,6 +136,30 @@ object BlueprintCommand : SLCommand() {
...
 	}
 
+	private fun calculateBlueprintCost(blueprint: Blueprint) : Double {
+		val clipboard = blueprint.loadClipboard()
+
+		var PriceAmmount = 0.0
+
+		val materialmap = mutableMapOf<BlockData, Int>()
+
+		clipboard.region.map {it}.forEach {
```
Comment (Peter-Crawley) [inline src/main/kotlin/net/starlegacy/command/starship/BlueprintCommand.kt:146]: i'll clean this up later

**Review by Peter-Crawley — CHANGES_REQUESTED** [review]

**Comment by Peter-Crawley** [comment]
has this been tested?

**Comment by SciathGH** [comment]
extensively
well that is before the clipboard.region.map{it}.foreach{} change but I dont see a reason why that would make a difference

**Comment by SciathGH** [comment]
> has this been tested?

And also it wont break severely, since I last tested only the blueprint info could possibly break Which is both an easy fix, and not a detriment

**Pull request merged**